### DISCUSSION TITLE
Multiple sinks with different output formats

### DISCUSSION THREAD

**Discussion by nstinus** [tokio-rs/tracing#1504] [Q&A] [answered]
Hi,

I am using tracing-subscriber 0.2.19.
I'd like to get several writers (that is covered in the examples). What I was unable to understand was how I can configure a different output format for each of them. I'd like to be able to get:
- to stdout: full format with max_level at WARN;
- to a rolling file, full format with a different max_level, spans activated (ENTER | CLOSE).

Q1: Is there a way to to this?
Q2: Ultimately, I'd like to file to have a different format like: timestamp_rfc3339 thread_name level <msg> <fields...> [source:line]

Thanks!

**Comment by hawkw** [comment, marked as answer]
This can be accomplished with [output filtering](https://docs.rs/tracing-subscriber/0.2.19/tracing_subscriber/fmt/writer/trait.MakeWriterExt.html). To also have separate formats, you can construct two `fmt` layers, using a [custom event formatter](https://docs.rs/tracing-subscriber/0.2.19/tracing_subscriber/fmt/struct.Layer.html#method.event_format) for one, and giving each one a writer with a different output format.

Hope that helps --- let me know if you have additional questions?

**Reply by nstinus** [reply]
Hi,

Thank you. I finally got it working:
```
    let file_appender = RollingFileAppender::new(
        Rotation::DAILY,
        std::env::var("LOG_DIR").unwrap_or_else(|_| "/tmp".to_string()),
        format!("{}.log", APP_NAME),
    );

    let (file_appender, _guard) = tracing_appender::non_blocking(file_appender);
    let file_appender = file_appender.with_max_level(Level::TRACE);
    let file_layer = Layer::new()
        .with_writer(file_appender)
        .with_thread_names(true)
        .with_span_events(FmtSpan::ENTER | FmtSpan::CLOSE)
        .with_timer(ChronoLocal::rfc3339())
        .with_ansi(false)
        .event_format(Formatter {});

    let (stdout_writer, _guard) = tracing_appender::non_blocking(std::io::stdout());
    let stdout_writer = stdout_writer.with_max_level(Level::WARN);
    let stdout_layer = Layer::new()
        .compact()
        .with_writer(stdout_writer)
        .with_timer(ChronoLocal::with_format("%F %T".to_string()));

    let subscriber = tracing_subscriber::registry()
        .with(file_layer)
        .with(stdout_layer);

    tracing::subscriber::set_global_default(subscriber).expect("Unable to set a global collector");
```
Now I have a few follow-up questions:
Q1: is this the idiomatic way to do it?
Q2: since it wasn't obvious to me, perhaps there is an interest in adding this in the examples? I would be happy to submit a PR.

Thanks!

**Reply by davidbarsky** [reply]
> Q1: is this the idiomatic way to do it?

Yup! This seems reasonable to me.

> Q2: since it wasn't obvious to me, perhaps there is an interest in adding this in the examples? I would be happy to submit a PR.

There is interest! I'd suggest creating a new example along the lines of [`fmt-multiple-writers.rs`](https://github.com/tokio-rs/tracing/blob/master/examples/examples/fmt-multiple-writers.rs) named `fmt-output-filtering.rs`.
